algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+        algorithmFPType sumN1N2    = *mrgLN + n;
+        algorithmFPType mulN1N2    = *mrgLN * n;
+        algorithmFPType deltaScale = mulN1N2 / sumN1N2;
+        algorithmFPType meanScale  = (algorithmFPType)1 / sumN1N2;
+        algorithmFPType delta      = mean - *mrgLMean;
+
+        *mrgLSum2Cent = *mrgLSum2Cent + sum2Cent + delta * delta * deltaScale;
+        *mrgLMean     = (*mrgLMean * *mrgLN + mean * n) * meanScale;
+        *mrgLN        = sumN1N2;
+    }
+
+    void mergeValToStat(algorithmFPType val, algorithmFPType * mrgN, algorithmFPType * mrgMean, algorithmFPType * mrgSum2Cent) {
+        *mrgN += (algorithmFPType)1;
+        algorithmFPType invN  = ((algorithmFPType)1) / *mrgN;
+        algorithmFPType delta = val - *mrgMean;
+        *mrgMean += delta * invN;
+        *mrgSum2Cent += delta * (val - *mrgMean);
+    }
+
+    __kernel void computeBestSplit(const __global int * data, const __global int * treeOrder, const __global int * selectedFeatures,
+                                   int nSelectedFeatures, const __global algorithmFPType * response, const __global int * binOffsets,
+                                   __global int * nodeList, const __global int * nodeIndices, int nodeIndicesOffset,
+                                   __global algorithmFPType * splitInfo, __global algorithmFPType * nodeImpDecreaseList,
+                                   int updateImpDecreaseRequired, int nFeatures, int minObservationsInLeafNode, algorithmFPType impurityThreshold) {
+        // this kernel is targeted for processing nodes with small number of rows
+        // nodeList will be updated with split attributes
+        // spliInfo will contain node impurity and mean
+        const int nNodeProp = 5; // num of node properties in nodeList
+        const int nImpProp  = 2;
+
+        const int local_id           = get_local_id(0);
+        const int sub_group_local_id = get_sub_group_local_id();
+        const int sub_group_size     = get_sub_group_size();
+        const int local_size         = get_local_size(0);
+        const int n_sub_groups       = local_size / sub_group_size; // num of subgroups for current node processing
+        const int sub_group_id       = local_id / sub_group_size;
+        const int max_sub_groups_num = 16; //replace with define
+
+        const int bufIdx  = get_global_id(1) % (max_sub_groups_num / n_sub_groups); // local buffer is shared between 16 sub groups
+        const int nodeIdx = get_global_id(1);
+        const int nodeId  = nodeIndices[nodeIndicesOffset + nodeIdx];
+
+        const int rowsOffset = nodeList[nodeId * nNodeProp + 0];
+        const int nRows      = nodeList[nodeId * nNodeProp + 1];
+
+        // each sub group will process 16 bins and produce 1 best split for it
+        // expect maximum 16 subgroups only for each node
+        const int maxBinsBlocks = 16;
+        __local algorithmFPType bufI[maxBinsBlocks]; // storage for impurity decrease
```

Comment:
Array with size <= 32 should be in private memory (in registers). It may be much faster

Reply:
__local array is used only in case when more than one subgroup is used for node processing. The goal of this array is making a final best split selection between results from all subgroups targeted for this node, do you see a way how private memory can be used for this?

Reply:
Agree